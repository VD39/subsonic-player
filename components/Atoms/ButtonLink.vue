<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | Icon;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: IconSize;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 'medium',
    iconWeight: 'regular',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        'centerAll fullWidth': fullWidth,
      },
    ]"
    :disabled
    draggable="false"
    :type
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      aria-hidden="true"
      :class="$style.icon"
      :color="iconColor"
      :size="ICON_SIZE[iconSize]"
      :weight="iconWeight"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
