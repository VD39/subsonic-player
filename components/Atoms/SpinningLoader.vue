<script setup lang="ts">
withDefaults(
  defineProps<{
    size?: number;
  }>(),
  {
    size: ICON_SIZE.medium,
  },
);
</script>

<template>
  <span :class="['centerItems', $style.spinningLoader]">
    <component
      :is="ICONS.spinner"
      ref="spinnerIcon"
      :size="size"
      :class="$style.icon"
    />

    <span class="visuallyHidden">Loading...</span>
  </span>
</template>

<style module>
.spinningLoader {
  flex-shrink: 0;
}

.icon {
  animation: spinner-loader 0.75s linear infinite;
}

@keyframes spinner-loader {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}
</style>
