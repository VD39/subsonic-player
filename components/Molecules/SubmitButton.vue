<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import SpinningLoader from '@/components/Atoms/SpinningLoader.vue';

const props = defineProps<{
  fullWidth?: boolean;
  loading?: boolean;
}>();

const icon = computed(() => (props.loading ? SpinningLoader : undefined));
</script>

<template>
  <ButtonLink
    :class="$style.submitButton"
    :disabled="loading"
    :fullWidth
    :icon
    iconPosition="right"
    showText
    type="submit"
  >
    <slot />
  </ButtonLink>
</template>

<style module>
.submitButton {
  --button-hover-color: var(--body-background-color);

  background-color: var(--button-hover-color);
  border: 1px solid var(--theme-color);
  border-radius: var(--border-radius-medium);

  @media (hover: hover) {
    &:hover,
    &:focus {
      --button-hover-color: var(--hover-selected-color);
    }
  }
}
</style>
