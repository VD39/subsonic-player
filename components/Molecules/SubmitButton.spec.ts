import type { VueWrapper } from '@vue/test-utils';

import { mount } from '@vue/test-utils';

import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import SpinningLoader from '@/components/Atoms/SpinningLoader.vue';

import SubmitButton from './SubmitButton.vue';

function factory(props = {}) {
  return mount(SubmitButton, {
    props: {
      ...props,
    },
    slots: {
      default: 'Default slot content.',
    },
  });
}

describe('SubmitButton', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when fullWidth prop is not set', () => {
    it('sets the correct ButtonLink fullWidth prop', () => {
      expect(wrapper.findComponent(ButtonLink).props('fullWidth')).toBe(false);
    });
  });

  describe('when fullWidth prop is set to true', () => {
    beforeEach(() => {
      wrapper = factory({
        fullWidth: true,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('sets the correct ButtonLink fullWidth prop', () => {
      expect(wrapper.findComponent(ButtonLink).props('fullWidth')).toBe(true);
    });
  });

  describe('when loading prop is not set', () => {
    it('sets the correct ButtonLink icon prop', () => {
      expect(wrapper.findComponent(ButtonLink).props('icon')).not.toBeDefined();
    });

    it('sets the correct ButtonLink disabled prop', () => {
      expect(wrapper.findComponent(ButtonLink).props('disabled')).toBe(false);
    });
  });

  describe('when loading prop is set to true', () => {
    beforeEach(() => {
      wrapper = factory({
        loading: true,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('sets the correct ButtonLink icon prop', () => {
      expect(wrapper.findComponent(ButtonLink).props('icon')).toBe(
        SpinningLoader,
      );
    });

    it('sets the correct ButtonLink disabled prop', () => {
      expect(wrapper.findComponent(ButtonLink).props('disabled')).toBe(true);
    });
  });
});
