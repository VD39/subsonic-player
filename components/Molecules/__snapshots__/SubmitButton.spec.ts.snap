// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubmitButton > matches the snapshot 1`] = `
"<button class="centerItems buttonLink submitButton submitButton" draggable="false" type="submit">
  <!--v-if--><span class="clamp text">Default slot content.</span>
</button>"
`;

exports[`SubmitButton > when fullWidth prop is set to true > matches the snapshot 1`] = `
"<button class="centerItems buttonLink centerAll fullWidth submitButton submitButton" draggable="false" type="submit">
  <!--v-if--><span class="clamp text">Default slot content.</span>
</button>"
`;

exports[`SubmitButton > when loading prop is set to true > matches the snapshot 1`] = `"<button class="centerItems buttonLink alignRight disabled submitButton submitButton" disabled="" draggable="false" type="submit"><span class="centerItems spinningLoader icon" aria-hidden="true" weight="regular"><phcirclenotch class="icon" size="24"></phcirclenotch><span class="visuallyHidden">Loading...</span></span><span class="clamp text">Default slot content.</span></button>"`;
